src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBot.cs

```diff
@@ -0,0 +1,623 @@
+                for (int i = 0, n = ranges.Length; i < n; ++i)
+                {
+                    var val = ranges[i].MidPoint;
+                    yield return val;
+                }
+            }
+        }
+    
+        /// <summary>
+        /// Generates a discrete action space for the game.
+        /// </summary>
+        protected virtual List<QAction> GenerateActionSpace()
+        {
+            var actionSpace = new List<QAction>();
+            foreach (var act in RGActionManager.Actions)
+            {
+                if (!RGActionManager.IsActionEnabled(act))
+                {
+                    continue;
+                }
+    
+                foreach (var param in GetDiscreteParamValues(act))
+                {
+                    actionSpace.Add(new QAction(act, param));
+                }
+            }
+            return actionSpace;
+        }
+
+        /// <summary>
+        /// Bots can define their own custom episode termination condition (besides EpisodeDuration) by overriding this method.
+        /// </summary>
+        protected virtual bool IsEpisodeFinished()
+        {
+            return false;
+        }
+    
+        private void SaveModel()
+        {
+            JObject result = new JObject();
+            result["epsilon"] = _epsilon;
+            JObject qtable = new JObject();
+            foreach (var stateEntry in _qTable)
+            {
+                JObject actionVals = new JObject();
+                foreach (var actionEntry in stateEntry.Value)
+                {
+                    actionVals[actionEntry.Key] = actionEntry.Value;
+                }
+                qtable[stateEntry.Key] = actionVals;
+            }
+            result["qtable"] = qtable;
+            File.WriteAllText(ModelFilePath, result.ToString());
+        }
+    
+        private bool LoadModel()
+        {
+            if (!File.Exists(ModelFilePath))
+            {
+                return false;
+            }
+            JObject data = JObject.Parse(System.IO.File.ReadAllText(ModelFilePath));
+            _epsilon = data["epsilon"].ToObject<float>();
+            JObject qtable = (JObject)data["qtable"];
+            _qTable = new Dictionary<string, Dictionary<string, float>>();
+            foreach (var entry in qtable)
+            {
+                Dictionary<string, float> actionVals = new Dictionary<string, float>();
+                JObject actionValsObj = (JObject)entry.Value;
+                foreach (var actionEntry in actionValsObj)
+                {
+                    actionVals[actionEntry.Key] = actionEntry.Value.ToObject<float>();
+                }
+                _qTable.Add(entry.Key, actionVals);
+            }
+            return true;
+        }
+
+        public void Start()
+        {
+            if (!RGActionManager.IsAvailable)
+            {
+                RGDebug.LogError("Action manager is currently unavailable");
```

Comment:
This caught me up in one of my early times using the framework. Mind if we add this?
```suggestion
                RGDebug.LogError("Action manager is currently unavailable. Have you run Regression Games > Analyze Actions on your project?");
```

Reply:
That does look more helpful! I went ahead and added that here and to the RGMonkeyBot as well.